pkgs/build-support/rust/build-rust-crate.nix

```diff
@@ -282,14 +282,48 @@ let makeDeps = dependencies:
 
          tr '\n' ' ' < target/link > target/link_
          LINK=$(cat target/link_)
-       fi
+      fi
 
       mkdir -p target/bin
-      echo "${crateBin}" | sed -n 1'p' | tr ',' '\n' | while read BIN; do
-         if [[ ! -z "$BIN" ]]; then
-           build_bin $BIN
-         fi
+      printf "%s\n" "${crateBin}" | head -n1 | tr -s ',' '\n' | while read -r BIN_NAME BIN_PATH; do
+        # filter empty entries / empty "lines"
+        if [[ -z "$BIN_NAME" ]]; then
+             continue
+        fi
+
+        if [[ -z "$BIN_PATH" ]]; then
+          # heuristic to "guess" the correct source file as found in cargo:
+          # https://github.com/rust-lang/cargo/blob/90fc9f620190d5fa3c80b0c8c65a1e1361e6b8ae/src/cargo/util/toml/targets.rs#L308-L325
+
+          # the first two cases are the "new" default IIRC
+          BIN_NAME_=$(echo $BIN_NAME | sed -e 's/_/-/g')
+          FILES="src/bin/$BIN_NAME_.rs src/bin/$BIN_NAME_/main.rs src/bin/main.rs src/main.rs"
+
+          if ! [ -e "${libPath}" -o -e src/lib.rs -o -e "src/${libName}.rs" ]; then
+            # if this is not a library the following path is also valid
+            FILES="src/$BIN_NAME_.rs $FILES"
+          fi
+
+          echo $FILES
+          for file in $FILES;
+          do
+            echo "checking file $file"
+            # first file that exists wins
+            if [[ -e "$file" ]]; then
+                    BIN_PATH="$file"
+                    break
+            fi
+          done
+
+          if [[ -z "$BIN_PATH" ]]; then
+            echo "failed to find file for binary target: $BIN_NAME" >&2
+            exit 1
+          fi
+        fi
+        build_bin $BIN_NAME $BIN_PATH
```

Comment:
Can these variables contain spaces?

Reply:
if they do I would guess the system build to fail in random places... I should add quotes there anyway I guess..

Reply:
the `buildRustCrate` expression support spaces now up to the point where the stdenv builder doesn't handle them well. I could add an extra level of escaping but the benefit is doubtful.